docs/proposals/31-inplace-node-update.md

```diff
@@ -0,0 +1,389 @@
+
+In the case of the `RollingUpdate` strategy, new machines are created in the new `MachineSet` using the new `MachineClass`, and the old ones are deleted from the old `MachineSet`. For in-place updates, instead of creating new machines for the new `MachineSet`, the old ones are moved from the old `MachineSet` to the new one after the `gardener-node-agent` has successfully updated the dependencies. This process preserves the old machines (their `Machine` and `Node` resources).
+
+There are two ways to achieve this:
+
+1. **Owner Reference Update**: Move the old `Machine` to the new `MachineSet` by updating the owner reference. The downside of this method is that the `Machine` name will stay the same through updates, which breaks the current naming convention of including the `MachineSet` name as a prefix in `Machine` name.
+2. **Shallow Delete**: Instead of directly moving a `Machine` from one `MachineSet` to another, first perform a shallow delete of the old `Machine` object (without deleting the underlying physical/virtual machine), then create a new `Machine` object in the new `MachineSet` (for the same physical/virtual machine).
+
+> [!NOTE]
+> At the time of writing this GEP, it is undecided which method to proceed with. The first method has been evaluated, but the feasibility of the second method is yet to be assessed. If feasible, the second method is preferred.
+
+#### `MachineDeployment` API
+
+A new strategy type `InPlaceUpdate` and a new field, `inPlaceUpdate`, have been introduced under the `.spec.strategy` field. `inPlaceUpdate` can be configured when the selected strategy type is `InPlaceUpdate`.
+
+```yaml
+apiVersion: machine.sapcloud.io/v1alpha1
+kind: MachineDeployment
+metadata:
+  name: <name>
+  namespace: <shoot-namespace>
+spec:
+  replicas: 1
+  selector:
+    matchLabels:
+      name: <label>
+  strategy:
+    inPlaceUpdate:
+      maxUnavailable: 1
+      maxSurge: 0
+      onLabel: false # true/false
+    type: InPlaceUpdate # RollingUpdate/InPlaceUpdate/Recreate
+  template: {}
+    spec: {}
+status: {}
+```
+
+`maxUnavailable`: Specifies the maximum number of unavailable machines during the update process.
+
+`maxSurge`: Specifies the maximum number of machines that can be scheduled above the desired number of machines. Set `maxSurge` to zero if you don't want any new machines/nodes to be provisioned during the update. You can set it to a value greater than zero if you don't want unavailable nodes during the update; in that case, first new machines/nodes will be created with the updated configuration and then the old machines/nodes will undergo in-place updates.
```

Comment:
Setting `maxSurge` to something greater than zero may lead to new machines being created and hence having more nodes than before the update. Therefore, some nodes may be scaled down later on.
Would this not somewhat reduce the effect of the in-place update as you may "loose" old machines over time? I am all for the flexibility, but I am a bit confused if this is something anybody would actually want.
WDYT?

Reply:
Just a comment: Whether `maxSurge` or out-scaling through the cluster-autoscaler, in-place updates with a drain (in contrast to the non-drain option that was rejected until now) may easily result in a situation where new nodes are necessary, because we remove capacity from the cluster during the drain/reboot and the pods must go somewhere then (unless they are node-sticky, but that will probably be the minority of the cases).

Of course, for manual bare metal clusters, `maxSurge` cannot be anything else than 0 and `minimum` must be equal to `maximum` to disable the cluster-autoscaler. Of course, we haven't yet defined how manual bare metal clusters will look like at spec level, but I imagine they must somewhat look like that. :-)